### PR TITLE
expression: support builtin function "benchmark"

### PR DESCRIPTION
**Pull request by vikingM**
### What problem does this PR solve? 
https://github.com/pingcap/tidb/issues/6774

### What is changed and how it works?
add the benchmark function
now this function args are (int,string)

### Check List 

Tests 

 - Manual test
select benchmark(10000,md5(rand())); 

Code changes
in  expression/builtin_info.go
add the related codes

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8313)

### CODE REVIEW

File: expression/builtin_info.go [hunk 1]
```diff
@@ -384,7 +385,39 @@ type benchmarkFunctionClass struct {
 }
 
 func (c *benchmarkFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (builtinFunc, error) {
-	return nil, errFunctionNotExists.GenWithStackByArgs("FUNCTION", "BENCHMARK")
+	//return nil, errFunctionNotExists.GenWithStackByArgs("FUNCTION", "BENCHMARK")
+	if err := c.verifyArgs(args); err != nil {
+		return nil, errors.Trace(err)
+	}
+	bf := newBaseBuiltinFuncWithTp(ctx, args, types.ETInt, types.ETInt, types.ETString)
```
Comment (zz-jason) [inline expression/builtin_info.go]: the second parameter can be other types, for example, `benchmark(1000, 1+2)`

File: pre-commit [hunk 2]
```diff
@@ -0,0 +1 @@
+../../hooks/pre-commit
```
Comment (zz-jason) [inline pre-commit]: I think this file is irrelevant, could you remove it from this PR?

File: expression/builtin_info.go [hunk 3]
```diff
@@ -384,7 +385,39 @@ type benchmarkFunctionClass struct {
...
+
+func (b *builtinBenchmarkSig) evalInt(row chunk.Row) (int64, bool, error) {
+	x, isNull, err := b.args[0].EvalInt(b.ctx, row)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
+	}
+	var i int64 = 0
+	for ; i < x; i++ {
+		_, isNull, err := b.args[1].EvalString(b.ctx, row)
+		if isNull || err != nil {
```
Comment (zz-jason) [inline expression/builtin_info.go]: We only need to check `err`.

File: expression/builtin_info.go [hunk 4]
```diff
@@ -384,7 +385,39 @@ type benchmarkFunctionClass struct {
 }
 
 func (c *benchmarkFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (builtinFunc, error) {
-	return nil, errFunctionNotExists.GenWithStackByArgs("FUNCTION", "BENCHMARK")
+	//return nil, errFunctionNotExists.GenWithStackByArgs("FUNCTION", "BENCHMARK")
```
Comment (zz-jason) [inline expression/builtin_info.go]: please remove this line.

**Comment by zz-jason** [comment]
@vikingM Please fix the CI.

**Comment by vikingM** [comment]
the conflict file just because when i create a json object, i found in "datumsToConstants" think it is a string object

File: expression/builtin_info.go [hunk 5]
```diff
@@ -384,7 +391,212 @@ type benchmarkFunctionClass struct {
...
+func (b *builtinBenchmarkStringSig) evalInt(row chunk.Row) (int64, bool, error) {
+	x, isNull, err := b.args[0].EvalInt(b.ctx, row)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
+	}
+	var i int64 = 0
+	for ; i < x; i++ {
+		_, _, err := b.args[1].EvalString(b.ctx, row)
+		if err != nil {
+			return 0, isNull, errors.Trace(err)
```
Comment (zz-jason) [inline expression/builtin_info.go]: `return 0, true, errors.Trace(err)`, you may need to change the other places in this patch as well.

File: expression/builtin_info.go [hunk 6]
```diff
@@ -384,7 +391,212 @@ type benchmarkFunctionClass struct {
...
+	baseBuiltinFunc
+}
+
+func (b *builtinBenchmarkStringSig) evalInt(row chunk.Row) (int64, bool, error) {
+	x, isNull, err := b.args[0].EvalInt(b.ctx, row)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
+	}
+	var i int64 = 0
+	for ; i < x; i++ {
```
Comment (zz-jason) [inline expression/builtin_info.go]: can be simplified to:
```go
for i := int64(0); i < x; i++ {
```

File: expression/evaluator_test.go [hunk 7]
```diff
@@ -111,6 +116,7 @@ func (s *testEvaluatorSuite) datumsToConstants(datums []types.Datum) []Expressio
 	constants := make([]Expression, 0, len(datums))
 	for _, d := range datums {
 		ft := s.kindToFieldType(d.Kind())
+		log.Print(d.Kind())
```
Comment (zz-jason) [inline expression/evaluator_test.go]: remove the debug log.

File: expression/builtin_info_test.go [hunk 8]
```diff
@@ -14,18 +14,19 @@
...
 	"github.com/pingcap/parser/charset"
 	"github.com/pingcap/parser/mysql"
 	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/types/json"
 	"github.com/pingcap/tidb/util/chunk"
 	"github.com/pingcap/tidb/util/mock"
 	"github.com/pingcap/tidb/util/printer"
 	"github.com/pingcap/tidb/util/testleak"
+	"math"
+	"time"
```
Comment (zz-jason) [inline expression/builtin_info_test.go]: please reorg the imported packages to two sections:
```go
"math"
"time"

"github.com/...."
...
```

File: expression/builtin_info_test.go [hunk 9]
```diff
@@ -121,9 +122,35 @@ func (s *testEvaluatorSuite) TestVersion(c *C) {
...
+		{[]interface{}{1000, 2}, 0, false, false},
+		{[]interface{}{1000, 2.0}, 0, false, false},
+		{[]interface{}{1000, "test"}, 0, false, false},
+		{[]interface{}{1000, types.TimeFromDays(20171010123456)}, 0, false, false},
+		{[]interface{}{1000, types.NewDecFromFloatForTest(1.1)}, 0, false, false},
+		{[]interface{}{1000, types.Duration{Duration: time.Duration(0)}}, 0, false, false},
+		{[]interface{}{1000, jsonObj}, 0, false, false},
+	}
+	for _, test := range cases {
+
```
Comment (zz-jason) [inline expression/builtin_info_test.go]: remove this empty line.

File: expression/builtin_info_test.go [hunk 10]
```diff
@@ -121,9 +122,35 @@ func (s *testEvaluatorSuite) TestVersion(c *C) {
...
+		f, err := fc.getFunction(s.ctx, s.datumsToConstants(types.MakeDatums(test.args...)))
+
+		c.Assert(err, IsNil)
+		v, err := evalBuiltinFunc(f, chunk.Row{})
+		c.Assert(err, IsNil)
+		c.Assert(v.GetInt64(), Equals, test.expected)
+	}
+
```
Comment (zz-jason) [inline expression/builtin_info_test.go]: ditto

**Comment by zz-jason** [comment]
Please merge the latest master branch and resolve conflicts.

File: expression/builtin_info_test.go (hunk 10 above)
Comment (zz-jason) [reply expression/builtin_info_test.go]: @vikingM please remove this useless empty line.

File: expression/builtin_info.go [hunk 11]
```diff
@@ -384,7 +391,211 @@ type benchmarkFunctionClass struct {
...
+	case types.ETDatetime:
+		sig = &builtinBenchmarkTimeSig{bf}
+	case types.ETTimestamp:
+		sig = &builtinBenchmarkTimeSig{bf}
+	default:
+		return nil, errors.New("FUNCTION benchmark has wrong arg2")
+	}
+
+	return sig, nil
+
```
Comment (zz-jason) [inline expression/builtin_info.go]: this empty line can be removed.

**Comment by zz-jason** [comment]
@vikingM Please merge the latest master and resolve conflicts.

**Comment by zz-jason** [comment]
@vikingM any update? there is a conflict between this and the master branch. Could you update to the latest master and resolve the conflicts in your spare time?

**Comment by vikingM** [comment]
So sorry. I'll resolve the conflict before Friday.

Zhang Jian <notifications@github.com> 于2019年1月8日周二 下午2:42写道：

File: expression/builtin_info_test.go [hunk 12]
```diff
@@ -121,9 +123,32 @@ func (s *testEvaluatorSuite) TestVersion(c *C) {
...
+		getErr   bool
+	}{
+		{[]interface{}{1000, 2}, 0, false, false},
+		{[]interface{}{1000, 2.0}, 0, false, false},
+		{[]interface{}{1000, "test"}, 0, false, false},
+		{[]interface{}{1000, types.TimeFromDays(20171010123456)}, 0, false, false},
+		{[]interface{}{1000, types.NewDecFromFloatForTest(1.1)}, 0, false, false},
+		{[]interface{}{1000, types.Duration{Duration: time.Duration(0)}}, 0, false, false},
+		{[]interface{}{1000, jsonObj}, 0, false, false},
+	}
```
Comment (qw4990) [inline expression/builtin_info_test.go]: Could you please add some cases that generate some errors or nil results~?

**Comment by zz-jason** [comment]
@vikingM I'm sorry to inform you that builtin function `benchmark()` has been supported in this PR https://github.com/pingcap/tidb/pull/9252, I'm going to close this PR now. Anyway, thanks for your contribution.

**Pull request closed without merging**